### DISCUSSION TITLE
copilot.vim: `:Copilot disable` doesn't disable copilot?

### DISCUSSION THREAD

**Discussion by msbit** [community/community#57887] [Copilot Conversations]
### Select Topic Area

Bug

### Body

If I edit a file, running:

```
:Copilot status
```

shows me:

```
Copilot: Enabled and online
```

and then running:

```
:Copilot disable
:Copilot status
```

shows me:

```
Copilot: Disabled globally by :Copilot disable
```

But after closing vim, opening another file and running:

``` 
:Copilot status
```

again, I get:

```
Copilot: Enabled and online
```

and, indeed, I do receive inline suggestions as I type.

My expectation is that `:Copilot disable` would keep copilot disabled (ie not parsing files or providing suggestions) until I re-run `:Copilot enable` either in this vim session or another.

Versions:

```
Vim 9.0.0276
copilot/dist/agent.js 1.84.0
Node.js 16.20.0
```

**Comment by msbit** [comment]
Any movement on this one? I'm finding it very annoying as I spend time on different codebases which have different policies regarding this kind of code access, and the only option I have at the moment is to:

```
:Copilot signout
```

when switching to the more restrictive codebase.

**Comment by mike-clark-8192** [comment]
I'm not a copilot.vim expert or contributor -- I'm just another end-user, so take this with a grain of salt. In particular, I'm not 100% certain I've gotten the autocmd events covering all possible buffer scenarios. I bet someone will eventually chime in if I've missed something.

For my workflow, I wanted:

1. Copilot is always off by default
2. Explicitly enable copilot when working on a certain file
3. I do not want to sign out and sign in constantly

### This is what I tried first
##### Regarding item 1
I added `let g:copilot_enabled = 0` to my vimrc.

##### Regarding item 2:
When I'm working on a file where I'd like suggestions, this seems to work:
`:Copilot enable`
OR
`:let g:copilot_enabled = 1`

#### Caveat Emptor
Because [g]vim can use RPC when opening files to reuse existing VIM processes when opening new files, this could cause surprise autocompletions when you reuse an existing GUI window to open some unrelated files. Or, if you like having multiple buffers open or use vim-native window splits, I think this will enable it for everything you have open in that vim process, which is certainly not in the spirit of what I'm trying to achieve.

### Take two
My next idea is 
1. Leave copilot enabled globally.  (This is the default behavior after loading the plugin, so that's easy enough.)
2. Use `:autocmd` to aggressively automatically disable it (`let b:copilot_enabled = 0`) whenever any buffer is instantiated in vim.
3. Enable it explicitly by hand-typing the ex command `:let b:copilot_enabled = 1` when working on a particular file.


```vim
" autocmd that executes for all new buffers / when opening a file of
" any kind. This way I can leave copilot enabled globally,
" but have copilot disabled for every buffer by default, as if it were disabled globally.
au BufNewFile,BufRead * let b:copilot_enabled = 0
```

Now when you want to enable copilot for a particular buffer (which usually means "for a particular open file"), run this ex (colon) command:
`:let b:copilot_enabled = 1`. Even if you reuse this vim process (`:split` or `:new` or on Windows: right-click "Edit with existing vim"), the setting should still only ever apply to the one file (buffer) you had in focus when you executed `:let b:copilot_enabled = 1`. 

After executing this command, you should start seeing completions right away after resuming editing the file (e.g. going back into insert mode). As of the current release version of `copilot.vim`, the only reason you would not start seeing completions after doing this is if you have copilot disabled globally, you are signed out, or your plugin is broken. Or there's a ghost in your machine.

### Extra credit

I'm satisfied with disabled-everywhere, enable explicitly and by hand on a per-buffer basis, but I can offer a few suggestions for people who might want just a little more to happen automatically:

vim `autocmd`s are very powerful. You could for example conditionally automatically enable copilot based on filetype, path patterns (glob/regex), etc. If you were very adventurous you could use an autocmd to look in a file's directory for a hypothetical `.copilot` settings file that contains directory-/project-/workspace-specific copilot settings. You might even be able to use a modeline to store a copilot enablement setting inside each file individually. I don't really like modelines because I think they can be a little bit of a security risk, but I do like the idea of looking for a `.copilot` settings file kind of like how git searches for `.gitignore` files. 

But I think the simplest example of having slightly more automatic rules would be to use path patterns with `autocmd`:

##### Snippets

```vim
" I want to leave copilot on for all files except files in a certain directory:
au BufNewFile,BufRead * let b:copilot_enabled = 1
au BufNewFile,BufRead $HOME/corporate_ip/* let b:copilot_enabled = 0
```

```vim
" I want to leave copilot off for files except files in a certain directory:
au BufNewFile,BufRead * let b:copilot_enabled = 0
au BufNewFile,BufRead $HOME/foss/* let b:copilot_enabled = 1
```

```vim
" `:let b:copilot_enabled = 1` is hard to remember and annoying to type?
command ThisIsYourCaptainSpeaking let b:copilot_enabled = 1
" Invoke with `:YourCommandName` (`:ThisIsYourCaptainSpeaking` in this example.)
```

Remember to keep smiling and enjoy every day. These days it seems the [singularity ](https://en.wikipedia.org/wiki/Technological_singularity) might be just around the bend. [😅](https://www.youtube.com/watch?v=Si_mGxIzHlU)

**Reply by zuyu** [reply]
Copied from https://github.com/orgs/community/discussions/7191#discussioncomment-3109345:

```
" Disable Copilot by default
let g:copilot_enabled = v:false
```

is simpler than `au BufNewFile,BufRead * let b:copilot_enabled = 0`

**Comment by parkerbedlan** [comment]
I also wanted to disable copilot by default in my neovim setup, so I added this to my `init.lua` (after my `packer.lua` already ran):
```lua
vim.cmd [[packadd copilot.vim]]
vim.cmd(':Copilot disable')
```

**Comment by ubuntupunk** [comment]
I don't have copilot.vim, the service appears to have been drawn as  dependency into my lazyvim setup somehow, and it is aggressively demanding authorisation, even after issuing a copilot disable. Yo can remove it from LazyVim extras: http://www.lazyvim.org/extras/coding/copilot
